### PR TITLE
AP-2556 AP-2566

### PR DESCRIPTION
**Pull request by chiichang**

### CODE REVIEW

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java [hunk 1]
```diff
@@ -189,125 +189,139 @@ public DefaultChartDataCollection(APMLog apmLog) {
...
-            avgPTUnit = (last - first) / MAX_UNIT;
-            double current = first - avgPTUnit;
-            for(int i=0; i <= MAX_UNIT; i++) {
-                if(i == MAX_UNIT) current = last;
-                else current += avgPTUnit;
-                avgPTMap.put(current, 0);
+        if (allAverageProcessingTimes.size() > 0) {
+            if (interval > 0) {
+                Collections.sort(allAverageProcessingTimes);
+                double first = allAverageProcessingTimes.get(0);
```
Comment (nolantellis) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:215]: Duplicate Code. Can use. a template method, which does all the above process and then just.
1 liner to do average

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java [hunk 2]
```diff
@@ -189,125 +189,139 @@ public DefaultChartDataCollection(APMLog apmLog) {
...
-            double current = first - maxPTUnit;
-            for(int i=0; i <= MAX_UNIT; i++) {
-                if(i == MAX_UNIT) current = last;
-                else current += maxPTUnit;
-                maxPTMap.put(current, 0);
+        if (allMaxProcessingTimes.size() > 0) {
+            if (interval > 0) {
+                Collections.sort(allMaxProcessingTimes);
+                double first = allMaxProcessingTimes.get(0);
+                double last = allMaxProcessingTimes.get(allMaxProcessingTimes.size() - 1);
```
Comment (nolantellis) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:235]: Duplicate code

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java [hunk 3]
```diff
@@ -189,125 +189,139 @@ public DefaultChartDataCollection(APMLog apmLog) {
...
-            for(int i=0; i <= numOfUnits; i++) {
-                if(i == numOfUnits) current = last;
-                else current += ttlWTUnit;
-                ttlWTMap.put(current, 0);
+        if (allTotalWaitingTimes.size() > 0) {
+            if (interval > 0) {
+                Collections.sort(allTotalWaitingTimes);
+                int numOfUnits = MAX_UNIT;
+                double first = allTotalWaitingTimes.get(0);
+                double last = allTotalWaitingTimes.get(allTotalWaitingTimes.size() - 1);
```
Comment (nolantellis) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:255]: Duplicate code. Let me know if you would like me to make the change

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java [hunk 4]
```diff
@@ -189,125 +189,139 @@ public DefaultChartDataCollection(APMLog apmLog) {
...
-            for(int i=0; i <= numOfUnits; i++) {
-                if(i == numOfUnits) current = last;
-                else current += avgWTUnit;
-                avgWTMap.put(current, 0);
+        if (allAverageWaitingTimes.size() > 0) {
+            if (interval > 0) {
+                Collections.sort(allAverageWaitingTimes);
+                int numOfUnits = MAX_UNIT;
+                double first = allAverageWaitingTimes.get(0);
+                double last = allAverageWaitingTimes.get(allAverageWaitingTimes.size() - 1);
```
Comment (nolantellis) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:275]: Same as above

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java [hunk 5]
```diff
@@ -189,125 +189,139 @@ public DefaultChartDataCollection(APMLog apmLog) {
...
-                else current += maxWTUnit;
-                maxWTMap.put(current, 0);
+        if (allMaxWaitingTimes.size() > 0 ) {
+            if (interval > 0) {
+                Collections.sort(allMaxWaitingTimes);
+                int numOfUnits = MAX_UNIT;
+                double first = allMaxWaitingTimes.get(0);
+                double last = allMaxWaitingTimes.get(allMaxWaitingTimes.size() - 1);
+                maxWTUnit = (last - first) / numOfUnits;
+                double current = first - maxWTUnit;
```
Comment (nolantellis) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:297]: Same as above

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java [hunk 6]
```diff
@@ -189,125 +189,139 @@ public DefaultChartDataCollection(APMLog apmLog) {
...
+                double first = allCaseUtils.get(0);
+                double last = allCaseUtils.get(allCaseUtils.size() - 1);
+                caseUtilUnit = (last - first) / numOfUnits;
+                double current = first - caseUtilUnit;
+                for (int i = 0; i < numOfUnits; i++) {
+                    current += caseUtilUnit;
+                    caseUtilMap.put(current, 0);
+                }
+                caseUtilMap.put(last, 0);
+                caseUtilMap.put(last, 0);
```
Comment (nolantellis) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:323]: I think this is an extra line

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java [hunk 7]
```diff
@@ -189,125 +189,139 @@ public DefaultChartDataCollection(APMLog apmLog) {
...
-            double current = first - caseUtilUnit;
-            for(int i=0; i < numOfUnits; i++) {
-                current += caseUtilUnit;
-                caseUtilMap.put(current, 0);
+        if (allCaseUtils.size() > 0 ) {
+            if (interval > 0) {
+                Collections.sort(allCaseUtils);
+                int numOfUnits = (MAX_UNIT / 2);
+                double first = allCaseUtils.get(0);
+                double last = allCaseUtils.get(allCaseUtils.size() - 1);
```
Comment (nolantellis) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:315]: Same as above

**Review by nolantellis — CHANGES_REQUESTED** [review]

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java (hunk 7 above)
Comment (chiichang) [reply Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:315]: This class will be removed later. So, I'm not going to change the code.

Comment (nolantellis) [reply Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:315]: But this is new code that is added in the class. 
Il make the changes, till then you can work on other stuff in the same

**Review by raboczi — APPROVED** [review]
I made a minor tweak, removing the explicit version of the commons-math3 directory so that it inherits the same one the rest of the system is using.

I agree with Nolan's proposal for refactoring some of the duplicate code, but I propose approving the current PR for merging and making the refactoring a subsequent PR that Chii can approve.

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java (hunk 1 above)
Comment (nolantellis) [reply Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/DefaultChartDataCollection.java:215]: Will do this in separate task

**Review by nolantellis — APPROVED** [review]

**Pull request merged**